gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/Compare.java

```diff
@@ -167,6 +181,20 @@ public Compare negate() {
         }
     };
 
+    private static boolean bothAreNumber(Object first, Object second) {
+        return first instanceof Number && second instanceof Number;
+    }
+
+    private static boolean bothAreComparable(Object first, Object second) {
+        return first instanceof Comparable && second instanceof Comparable;
+    }
+
+    private static void checkBothAreNumberOrComparable(Object first, Object second) {
+        if (null != first && null != second && !bothAreNumber(first, second) && !bothAreComparable(first, second)) {
+            throw new IllegalArgumentException(String.format("'%s' and '%s' need to be an instance of Number or Comparable", first, second));
```

Comment:
I think that I like this approach better as the behavior is a bit more like  the old `ClassCastException`. The error message is good because it tells you what's expected and shows the actual values being compared. Perhaps it could be refined a bit further though. What do you think of this text:

```java
String.format("Cannot compare '%s' (%s) and '%s' (%s) as both need to be an instance of Number or Comparable", first, first.getClass().getSimpleName(), second, second.getClass().getSimpleName())
```

I like the idea of having the class names in there because that's what the old CCE used to have.  It's also a bit more clear that the error is a result of one or both of the values being a problem.

Reply:
You could do away with the double checking too...
```
   if (null != first && null != second) {
      if (bothAreNumber(first, second)) {
         return NumberHelper.compare((Number) first, (Number) second) < 0;
      } else if (bothAreComparable(first, second)) {
         return ((Comparable) first).compareTo(second) < 0;
      } else {
         throw new IllegalArgumentException(String.format("Cannot compare '%s' (%s) and '%s' (%s) as both need to be an instance of Number or Comparable", first, first.getClass().getSimpleName(), second, second.getClass().getSimpleName()));
      }
   } else 
      return false;
```

Reply:
addressed both of your comments and also added a check to make sure that both are actually of the same type (with tests)

Reply:
@robertdale @spmallette @dkuppitz 
can you please re-review?